File: lib/xnix/compatibility/runtime_service_binding.rb
# frozen_string_literal: true

require "json"
require "optparse"
require "pathname"

module Xnix
  module Compatibility
    class RuntimeServiceBinding
      PROJECT_ROOT = Pathname.new(__dir__).join("../../..").realpath
      VERSION = PROJECT_ROOT.join("VERSION").read.strip
      BUS_NAME = "org.xnix.Compatibility1"
      OBJECT_PATH = "/org/xnix/Compatibility1"
      INTERFACE = "org.xnix.Compatibility1"
      SERVICE_FILE = PROJECT_ROOT.join("runtime/dbus/org.xnix.Compatibility1.service")
      SYSTEMD_UNIT = PROJECT_ROOT.join("runtime/systemd/xnix-compatd.service")
      CONTRACT_FILE = PROJECT_ROOT.join("runtime/dbus/org.xnix.Compatibility1.xml")
      LIBEXEC_WRAPPER = PROJECT_ROOT.join("libexec/xnix/compatd")
      SMOKE_ADAPTER = PROJECT_ROOT.join("runtime/dbus/xnix_compatd_smoke.c")
      PACKAGED_WRAPPER = "/usr/libexec/xnix/compatd"

      def to_h
        {
          "version" => VERSION,
          "binding_type" => "runtime-service-binding",
          "runtime_owned" => true,
          "kde_policy_owner" => false,
          "bus_name" => BUS_NAME,
          "object_path" => OBJECT_PATH,
          "interface" => INTERFACE,
          "activation" => activation,
          "checks" => checks,
          "counts" => counts,
          "activation_binding_ready" => activation_binding_ready?,
          "live_dbus_owner_ready" => false,
          "smoke_adapter_available" => SMOKE_ADAPTER.file?,
          "network_required" => false,
          "host_root_modified" => false,
          "privileged_container_required" => false,
          "desktop_safe_summary" => desktop_safe_summary,
          "backend_details_exposed" => false
        }
      end

      def to_json(*args)
        JSON.pretty_generate(to_h, *args)
      end

      private

      def activation
        {
          "dbus_service_file" => relative(SERVICE_FILE),
          "systemd_unit" => relative(SYSTEMD_UNIT),
          "libexec_wrapper" => relative(LIBEXEC_WRAPPER),
          "dbus_contract" => relative(CONTRACT_FILE),
          "packaged_wrapper" => PACKAGED_WRAPPER
        }
      end

      def checks
        [
          dbus_service_check,
          systemd_unit_check,
          libexec_wrapper_check,
          contract_check,
          live_owner_check
        ]
      end

      def dbus_service_check
        content = read_optional(SERVICE_FILE)
        pass = content.include?("Name=#{BUS_NAME}") &&
               content.include?("SystemdService=xnix-compatd.service") &&
               content.include?("Exec=#{PACKAGED_WRAPPER}")
        check("dbus-service-activation", pass ? "pass" : "blocked", "D-Bus activation points to the packaged Runtime wrapper.")
      end

      def systemd_unit_check
        content = read_optional(SYSTEMD_UNIT)
        pass = content.include?("Type=dbus") &&
               content.include?("BusName=#{BUS_NAME}") &&
               content.include?("ExecStart=#{PACKAGED_WRAPPER}") &&
               content.include?("NoNewPrivileges=yes") &&
               content.include?("ProtectSystem=strict")
        check("systemd-service-hardening", pass ? "pass" : "blocked", "systemd activation uses the stable bus name and hardened service settings.")
      end

      def libexec_wrapper_check
        content = read_optional(LIBEXEC_WRAPPER)
        pass = LIBEXEC_WRAPPER.file? &&
               LIBEXEC_WRAPPER.executable? &&
               content.include?("runtime_daemon")
        check("libexec-wrapper", pass ? "pass" : "blocked", "Packaged Runtime wrapper delegates to the Runtime daemon entry point.")
      end

      def contract_check
        content = read_optional(CONTRACT_FILE)
        pass = content.include?(INTERFACE) &&
               content.include?("ListApplications") &&
               content.include?("GetRuntimeServiceBinding")
        check("dbus-contract", pass ? "pass" : "blocked", "D-Bus contract exposes read-only Runtime service binding status.")
      end

      def live_owner_check
        check("live-dbus-owner", "pending", "A long-running production D-Bus owner is still pending.")
      end

      def check(id, status, summary)
        {
          "id" => id,
          "status" => status,
          "summary" => summary
        }
      end

      def counts
        statuses = checks.map { |item| item.fetch("status") }
        {
          "total" => statuses.length,
          "passed" => statuses.count("pass"),
          "pending" => statuses.count("pending"),
          "blocked" => statuses.count("blocked")
        }
      end

      def activation_binding_ready?
        checks.reject { |item| item.fetch("id") == "live-dbus-owner" }.all? { |item| item.fetch("status") == "pass" }
      end

      def desktop_safe_summary
        return "Runtime service activation files are aligned; live D-Bus ownership remains pending." if activation_binding_ready?

        "Runtime service activation files need repair before production D-Bus ownership can be enabled."
      end

      def read_optional(path)
        return "" unless path.file?

        path.read
      end

      def relative(path)
        path.relative_path_from(PROJECT_ROOT).to_s
      end

      class CLI
        def initialize(argv)
          @argv = argv.dup
        end

        def run
          parser.parse!(@argv)
          puts RuntimeServiceBinding.new.to_json
          0
        rescue OptionParser::ParseError, ArgumentError => e
          warn "xnix-runtime-service-binding: #{e.message}"
          64
        end

        private

        def parser
          OptionParser.new do |options|
            options.banner = "Usage: xnix-runtime-service-binding"
          end
        end
      end
    end
  end
end

File: internal/runtime/appidentity/runtime_service_binding.go
package appidentity

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const (
	runtimeServiceBindingBusName          = "org.xnix.Compatibility1"
	runtimeServiceBindingObjectPath       = "/org/xnix/Compatibility1"
	runtimeServiceBindingInterface        = "org.xnix.Compatibility1"
	runtimeServiceBindingDBusServiceFile  = "runtime/dbus/org.xnix.Compatibility1.service"
	runtimeServiceBindingSystemdUnit      = "runtime/systemd/xnix-compatd.service"
	runtimeServiceBindingLibexecWrapper   = "libexec/xnix/compatd"
	runtimeServiceBindingDBusContract     = "runtime/dbus/org.xnix.Compatibility1.xml"
	runtimeServiceBindingSmokeAdapter     = "runtime/dbus/xnix_compatd_smoke.c"
	runtimeServiceBindingPackagedWrapper  = "/usr/libexec/xnix/compatd"
	runtimeServiceBindingProductionStatus = "pending-live-owner"
)

type RuntimeServiceBindingPreview struct {
	Version                     string                       `json:"version"`
	SchemaVersion               string                       `json:"schema_version"`
	RequestType                 string                       `json:"request_type"`
	BindingType                 string                       `json:"binding_type"`
	Source                      string                       `json:"source"`
	RuntimeMethod               string                       `json:"runtime_method"`
	ReadMethod                  string                       `json:"read_method"`
	ProductionStatus            string                       `json:"production_status"`
	BusName                     string                       `json:"bus_name"`
	ObjectPath                  string                       `json:"object_path"`
	Interface                   string                       `json:"interface"`
	Activation                  RuntimeServiceActivation     `json:"activation"`
	Checks                      []RuntimeServiceBindingCheck `json:"checks"`
	CheckIDs                    []string                     `json:"check_ids"`
	Counts                      RuntimeServiceBindingCounts  `json:"counts"`
	RuntimeOwned                bool                         `json:"runtime_owned"`
	GoRuntimeBacked             bool                         `json:"go_runtime_backed"`
	KDEPolicyOwner              bool                         `json:"kde_policy_owner"`
	ActivationBindingReady      bool                         `json:"activation_binding_ready"`
	LiveDBusOwnerReady          bool                         `json:"live_dbus_owner_ready"`
	SmokeAdapterAvailable       bool                         `json:"smoke_adapter_available"`
	ServiceStarted              bool                         `json:"service_started"`
	ProductionBusClaimed        bool                         `json:"production_bus_claimed"`
	KDEMayClaimRuntimeOwnership bool                         `json:"kde_may_claim_runtime_ownership"`
	NetworkRequired             bool                         `json:"network_required"`
	HostRootModified            bool                         `json:"host_root_modified"`
	PrivilegedContainerRequired bool                         `json:"privileged_container_required"`
	BackendDetailsExposed       bool                         `json:"backend_details_exposed"`
	BlockedActions              []string                     `json:"blocked_actions"`
	DesktopSafeSummary          string                       `json:"desktop_safe_summary"`
}

type RuntimeServiceActivation struct {
	DBusServiceFile string `json:"dbus_service_file"`
	SystemdUnit     string `json:"systemd_unit"`
	LibexecWrapper  string `json:"libexec_wrapper"`
	DBusContract    string `json:"dbus_contract"`
	PackagedWrapper string `json:"packaged_wrapper"`
}

type RuntimeServiceBindingCheck struct {
	ID      string `json:"id"`
	Status  string `json:"status"`
	Summary string `json:"summary"`
}

type RuntimeServiceBindingCounts struct {
	Total   int `json:"total"`
	Passed  int `json:"passed"`
	Pending int `json:"pending"`
	Blocked int `json:"blocked"`
}

func NewRuntimeServiceBindingPreview(root string) (RuntimeServiceBindingPreview, error) {
	if root == "" {
		root = "."
	}
	version, err := readRuntimeServiceBindingVersion(root)
	if err != nil {
		return RuntimeServiceBindingPreview{}, err
	}

	checks := []RuntimeServiceBindingCheck{
		runtimeServiceDBusActivationCheck(root),
		runtimeServiceSystemdHardeningCheck(root),
		runtimeServiceLibexecWrapperCheck(root),
		runtimeServiceContractCheck(root),
		runtimeServiceLiveOwnerCheck(),
	}
	counts := countRuntimeServiceBindingChecks(checks)
	activationReady := runtimeServiceActivationReady(checks)

	preview := RuntimeServiceBindingPreview{
		Version:          version,
		SchemaVersion:    "xnix.runtime.service_binding.v1",
		RequestType:      "runtime-service-binding-preview",
		BindingType:      "runtime-service-binding",
		Source:           "activation-files+dbus-contract+runtime-owner-gate",
		RuntimeMethod:    "GetRuntimeServiceBinding",
		ReadMethod:       "GetRuntimeServiceBindingPreview",
		ProductionStatus: runtimeServiceBindingProductionStatus,
		BusName:          runtimeServiceBindingBusName,
		ObjectPath:       runtimeServiceBindingObjectPath,
		Interface:        runtimeServiceBindingInterface,
		Activation: RuntimeServiceActivation{
			DBusServiceFile: runtimeServiceBindingDBusServiceFile,
			SystemdUnit:     runtimeServiceBindingSystemdUnit,
			LibexecWrapper:  runtimeServiceBindingLibexecWrapper,
			DBusContract:    runtimeServiceBindingDBusContract,
			PackagedWrapper: runtimeServiceBindingPackagedWrapper,
		},
		Checks:                      checks,
		CheckIDs:                    runtimeServiceBindingCheckIDs(checks),
		Counts:                      counts,
		RuntimeOwned:                true,
		GoRuntimeBacked:             true,
		KDEPolicyOwner:              false,
		ActivationBindingReady:      activationReady,
		LiveDBusOwnerReady:          false,
		SmokeAdapterAvailable:       runtimeServicePathExists(root, runtimeServiceBindingSmokeAdapter),
		ServiceStarted:              false,
		ProductionBusClaimed:        false,
		KDEMayClaimRuntimeOwnership: false,
		NetworkRequired:             false,
		HostRootModified:            false,
		PrivilegedContainerRequired: false,
		BackendDetailsExposed:       false,
		BlockedActions: []string{
			"start Runtime service from preview",
			"claim production D-Bus owner from preview",
			"let KDE claim Runtime ownership",
			"mutate host root during service binding planning",
		},
		DesktopSafeSummary: runtimeServiceBindingSummary(activationReady),
	}
	if err := validateNoBackendTerms(preview, "Runtime service binding preview"); err != nil {
		return RuntimeServiceBindingPreview{}, err
	}
	return preview, nil
}

func runtimeServiceDBusActivationCheck(root string) RuntimeServiceBindingCheck {
	content := readRuntimeServiceBindingOptional(root, runtimeServiceBindingDBusServiceFile)
	passed := strings.Contains(content, "Name="+runtimeServiceBindingBusName) &&
		strings.Contains(content, "SystemdService=xnix-compatd.service") &&
		strings.Contains(content, "Exec="+runtimeServiceBindingPackagedWrapper)
	return runtimeServiceBindingCheck("dbus-service-activation", runtimeServiceCheckStatus(passed), "D-Bus activation points to the packaged Runtime wrapper.")
}

func runtimeServiceSystemdHardeningCheck(root string) RuntimeServiceBindingCheck {
	content := readRuntimeServiceBindingOptional(root, runtimeServiceBindingSystemdUnit)
	passed := strings.Contains(content, "Type=dbus") &&
		strings.Contains(content, "BusName="+runtimeServiceBindingBusName) &&
		strings.Contains(content, "ExecStart="+runtimeServiceBindingPackagedWrapper) &&
		strings.Contains(content, "NoNewPrivileges=yes") &&
		strings.Contains(content, "ProtectSystem=strict")
	return runtimeServiceBindingCheck("systemd-service-hardening", runtimeServiceCheckStatus(passed), "systemd activation uses the stable bus name and hardened service settings.")
}

func runtimeServiceLibexecWrapperCheck(root string) RuntimeServiceBindingCheck {
	content := readRuntimeServiceBindingOptional(root, runtimeServiceBindingLibexecWrapper)
	info, err := os.Stat(runtimeServicePath(root, runtimeServiceBindingLibexecWrapper))
	passed := err == nil && !info.IsDir() && info.Mode().Perm()&0o111 != 0 && strings.Contains(content, "runtime_daemon")
	return runtimeServiceBindingCheck("libexec-wrapper", runtimeServiceCheckStatus(passed), "Packaged Runtime wrapper delegates to the Runtime daemon entry point.")
}

func runtimeServiceContractCheck(root string) RuntimeServiceBindingCheck {
	content := readRuntimeServiceBindingOptional(root, runtimeServiceBindingDBusContract)
	passed := strings.Contains(content, runtimeServiceBindingInterface) &&
		strings.Contains(content, "ListApplications") &&
		strings.Contains(content, "GetRuntimeServiceBinding")
	return runtimeServiceBindingCheck("dbus-contract", runtimeServiceCheckStatus(passed), "D-Bus contract exposes read-only Runtime service binding status.")
}

func runtimeServiceLiveOwnerCheck() RuntimeServiceBindingCheck {
	return runtimeServiceBindingCheck("live-dbus-owner", "pending", "A long-running production D-Bus owner is still pending.")
}

func runtimeServiceBindingCheck(id string, status string, summary string) RuntimeServiceBindingCheck {
	return RuntimeServiceBindingCheck{
		ID:      id,
		Status:  status,
		Summary: summary,
	}
}

func runtimeServiceCheckStatus(passed bool) string {
	if passed {
		return "pass"
	}
	return "blocked"
}

func countRuntimeServiceBindingChecks(checks []RuntimeServiceBindingCheck) RuntimeServiceBindingCounts {
	counts := RuntimeServiceBindingCounts{Total: len(checks)}
	for _, check := range checks {
		switch check.Status {
		case "pass":
			counts.Passed++
		case "pending":
			counts.Pending++
		case "blocked":
			counts.Blocked++
		}
	}
	return counts
}

func runtimeServiceBindingCheckIDs(checks []RuntimeServiceBindingCheck) []string {
	ids := make([]string, 0, len(checks))
	for _, check := range checks {
		ids = append(ids, check.ID)
	}
	return ids
}

func runtimeServiceActivationReady(checks []RuntimeServiceBindingCheck) bool {
	for _, check := range checks {
		if check.ID == "live-dbus-owner" {
			continue
		}
		if check.Status != "pass" {
			return false
		}
	}
	return true
}

func runtimeServiceBindingSummary(activationReady bool) string {
	if activationReady {
		return "Runtime service activation files are aligned; live D-Bus ownership remains pending."
	}
	return "Runtime service activation files need repair before production D-Bus ownership can be enabled."
}

func readRuntimeServiceBindingVersion(root string) (string, error) {
	content, err := os.ReadFile(runtimeServicePath(root, "VERSION"))
	if err != nil {
		return "", fmt.Errorf("read VERSION: %w", err)
	}
	version := strings.TrimSpace(string(content))
	if version == "" {
		return "", fmt.Errorf("read VERSION: empty version")
	}
	return version, nil
}

func readRuntimeServiceBindingOptional(root string, relativePath string) string {
	content, err := os.ReadFile(runtimeServicePath(root, relativePath))
	if err != nil {
		return ""
	}
	return string(content)
}

func runtimeServicePathExists(root string, relativePath string) bool {
	info, err := os.Stat(runtimeServicePath(root, relativePath))
	return err == nil && !info.IsDir()
}

func runtimeServicePath(root string, relativePath string) string {
	return filepath.Join(root, filepath.FromSlash(relativePath))
}
